firefoxThemes/resources.js

```diff
@@ -1,45 +1,239 @@
+        this.image = image;//Relative link to waifu image file
+        this.json = json;//Relative link to browser theme file
+        this.definition = themeDefinition;
+        this.page = "/waifus/index.html";
+    }
+}
+/*---FUNCTIONS---*/
+/*Initialize Local Storage & custom new tab page*/
+function startStorage() {
+    browser.storage.local.get("currentThemeId")
+        .then((storage) => {
+            const initStorage = {
+                waifuThemes: new WaifuThemes()
+            };
+            //Retrieve all themes if none exists in local storage
+            browser.storage.local.set(initStorage);
+            //Load browser theme
+            if (storage.currentThemeId) {
+                loadTheme(initStorage.waifuThemes.themes, storage.currentThemeId);
+            }
+            //When the browser first opens, redirect to custom new tab page
+            browser.tabs.update({loadReplace: true, url: "waifus/index.html"});
+        });
+}
+/*Update all new tabs with new waifu theme*/
+function updateTabs(msg) {
+    switch (msg.mixState){
+        case mixedStates.INITIAL:
+            SetupMixedUpdate();
+            break;
+        default:
+            MixTabCleanup();
+            NormalUpdate(msg);
+            break;
+    }
+}
+/*Initialize the Mixed feature*/
+function SetupMixedUpdate(){
+    browser.tabs.query({title:"New Tab"})
+        .then((tabs)=>{
+            browser.storage.local.get(["waifuThemes","mixedTabs"])
+                .then((storage)=>{
+                    if(!storage.mixedTabs){
+                        storage.mixedTabs = new Map();//Create a new mixed tab list
+                    }
+                    //Activate event listeners
+                    browser.tabs.onCreated.addListener(MixTabCreated);//When a tab is created
+                    browser.tabs.onActivated.addListener(MixTabActivated);//When the active tab has been switched
+                    browser.tabs.onRemoved.addListener(MixTabClosed);//When a tab has been closed
+                    if(tabs.length > 0){
+                        const themes = storage.waifuThemes.themes;
+                        let currentThemeId = getRandomTheme(themes);
+                        const lastTab = keepLastTab(tabs);//Closes all New Tab tabs except the last
+                        storage.mixedTabs.set(lastTab.id,currentThemeId);//Add a default theme to the mixed tab list
+                        browser.storage.local.set({currentThemeId:currentThemeId,mixedTabs:storage.mixedTabs});
+                        //Initialize each tab with the default waifu
+                        browser.tabs.update(lastTab.id, {
+                            loadReplace: true,
+                            url: themes[currentThemeId].page
+                        });
+                        loadTheme(themes,currentThemeId);
+                    }
+
+                });
+        });
+}
+/*Closes all but the last tab*/
+function keepLastTab(tabs){
```

Comment:
Was there a technical limitation that requires the extension to close all other tabs?

Reply:
Yes. Since all themes load the same HTML document, only the last background image added to the local storage can be loaded. Even when multiple messages are sent from the background script to each tab's HTML page, new changes will not be applied. Even with a reload.

So two solutions:
1. Close all _new tabs_ and start from beginning
2. Create one HTML document for each background image. Each background image will have its own dedicated HTML which will load it.

Reply:
Got it! I don't think it's that big of an issue. It just closes the empty tabs. So I'm not too concerned